java/ql/src/semmle/code/java/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1450,36 +1460,35 @@ private predicate contentStoreStep(
 }
 
 /**
- * Holds if data may flow from `mid` to an exit of `m` in the context
- * `innercc`, and the path did not flow through a parameter of `m`.
+ * Holds if data may flow from `mid` to a return at position `pos` in the
+ * context `innercc`, and the path did not flow through a parameter.
  */
-private predicate flowOutOfMethod0(PathNodeMid mid, Method m, CallContext innercc) {
-  exists(ReturnNode ret |
-    ret = mid.getNode() and
-    innercc = mid.getCallContext() and
-    m = returnNodeGetEnclosingCallable(ret) and
-    not innercc instanceof CallContextCall
-  )
+private predicate flowOutOfCallable0(PathNodeMid mid, ReturnPosition pos, CallContext innercc) {
+  pos.getAReturnNode() = mid.getNode() and
+  innercc = mid.getCallContext() and
+  not innercc instanceof CallContextCall
 }
 
 /**
- * Holds if data may flow from `mid` to `ma`. The last step of this path
- * is a return from a method and is recorded by `cc`, if needed.
+ * Holds if data may flow from `mid` to `out`. The last step of this path
+ * is a return from a callable and is recorded by `cc`, if needed.
  */
 pragma[noinline]
-private predicate flowOutOfMethod(PathNodeMid mid, MethodAccess ma, CallContext cc) {
-  exists(Method m, CallContext innercc |
-    flowOutOfMethod0(mid, m, innercc) and
-    resolveReturn(innercc, m, ma)
+private predicate flowOutOfCallable(PathNodeMid mid, Node out, CallContext cc) {
+  exists(ReturnPosition pos, DataFlowCallable c, DataFlowCall call, CallContext innercc |
+    flowOutOfCallable0(mid, pos, innercc) and
+    out = getAnOutputAtCall(call, pos) and
```

Comment:
Again this is dangerous. Two copies of the virtual dispatch relation are in play, when only one ought to be needed.

Reply:
This one is a bit different, because `resolveReturn()` takes call contexts into account, which `getAnOutputAtCall()` does not, so not sure how that should be refactored.

Reply:
We need to add a variant of `resolveReturn` that matches a `ReturnPosition` to an `OutNode`.